mutable_tree.go

```diff
@@ -848,55 +781,66 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 	if tree.VersionExists(version) {
 		// If the version already exists, return an error as we're attempting to overwrite.
 		// However, the same hash means idempotent (i.e. no-op).
-		existingHash, err := tree.ndb.getRoot(version)
+		existingNodeKey, err := tree.ndb.GetRoot(version)
 		if err != nil {
 			return nil, version, err
 		}
-
-		// If the existing root hash is empty (because the tree is empty), then we need to
-		// compare with the hash of an empty input which is what `WorkingHash()` returns.
-		if len(existingHash) == 0 {
-			existingHash = sha256.New().Sum(nil)
+		var existingRoot *Node
+		if existingNodeKey.version != 0 {
+			existingRoot, err = tree.ndb.GetNode(existingNodeKey)
+			if err != nil {
+				return nil, version, err
+			}
 		}
 
 		newHash, err := tree.WorkingHash()
 		if err != nil {
 			return nil, version, err
 		}
 
-		if bytes.Equal(existingHash, newHash) {
+		if (existingRoot == nil && tree.root == nil) || (existingRoot != nil && bytes.Equal(existingRoot.hash, newHash)) { // TODO with WorkingHash
 			tree.version = version
+			tree.root = existingRoot
 			tree.ImmutableTree = tree.ImmutableTree.clone()
 			tree.lastSaved = tree.ImmutableTree.clone()
-			tree.orphans = map[string]int64{}
-			return existingHash, version, nil
+			tree.orphans = []*NodeKey{}
+			return newHash, version, nil
 		}
 
-		return nil, version, fmt.Errorf("version %d was already saved to different hash %X (existing hash %X)", version, newHash, existingHash)
+		return nil, version, fmt.Errorf("version %d was already saved to different hash from %X (existing nodeKey %d)", version, newHash, existingNodeKey)
 	}
 
-	if tree.root == nil {
-		// There can still be orphans, for example if the root is the node being
-		// removed.
-		logger.Debug("SAVE EMPTY TREE %v\n", version)
-		if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
-			return nil, 0, err
-		}
-		if err := tree.ndb.SaveEmptyRoot(version); err != nil {
-			return nil, 0, err
-		}
-	} else {
-		logger.Debug("SAVE TREE %v\n", version)
-		if _, _, err := tree.ndb.SaveBranch(tree.root); err != nil {
+	logger.Debug("SAVE TREE %v\n", version)
+	// save orphans
+	if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
+		return nil, 0, err
+	}
+
+	// save new nodes
+	if tree.root != nil {
+		newNodes, err := tree.getNewNodes()
```

Comment:
Why do we have to get all these nodes in memory and then iterate them as a map a second time only to save?

After we iterate to assign nonces, there must be another in-order tree iteration to make sure that we save nodes in lexicographic order.

Reply:
the purpose of `in-order` is to calculate the node hash and then to make sure of lexicographic order, it stores all nodes in the map and iterates again in the sort order of nonce. I think we can't do both `hash calc` and `nonce assigning`.
The other purpose is I'd like to keep the nonce of the root as 1, and keep the ASC order of nodes when visiting the leaf node.

Reply:
What if we do this?

```go
	recursiveSpread = func(node *Node) (*NodeKey, error) {
		if node.nodeKey != nil {
			return node.nodeKey, nil
		}

		var err error
		if node.leftNode != nil {
			node.leftNodeKey, err = recursiveSpread(node.leftNode)
			if err != nil {
				return nil, err
			}
		}

		nonce++
		node.nodeKey = &NodeKey{
			version: version,
			nonce:   nonce,
		}

		if node.rightNode != nil {
			node.rightNodeKey, err = recursiveSpread(node.rightNode)
			if err != nil {
				return nil, err
			}
		}

		_, err = node._hash(version)
		if err != nil {
			return nil, err
		}

		node.leftNode = nil
		node.rightNode = nil
		newNodes[node.nodeKey.nonce] = node
		return node.nodeKey, nil
	}
```

This way, the nonce assignment follows in-order traversal while hashing follows post-order traversal as desired.

Reply:
Then, we can do a second in-order traversal of the tree to right the nodes to disk, removing the need for the map

Reply:
the second traversal makes sense, but I think the second traversal should be pre-order.